### PR TITLE
feat: Support nix package manager

### LINKED ISSUE #233: Support Nix package manager
We want to install barretenberg with the Nix package manager in Noir.

https://nixos.org/

### PR DESCRIPTION
**Pull request by phated**
# Description

Closes #233

This adds support for the Nix package manager and adds a GitHub Action to make sure the artifacts can be built.

~~Currently a draft because I'm trying to work through some difficulties with the Linux cross-compilation.~~

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: .github/workflows/nix.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+name: Nix builds
+
+on: [push]
```
Comment (phated) [inline .github/workflows/nix.yml]: I don't like this on push but I'm not sure what to restrict it to. Maybe just releases

File: barretenberg.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    # As per https://discourse.nixos.org/t/gcc11stdenv-and-clang/17734/7 since it seems that aarch64-linux uses
+    # gcc9 instead of gcc11 for the C++ stdlib, while all other targets we support provide the correct libstdc++
+      overrideCC llvmPackages.stdenv (llvmPackages.clang.override { gccForLibs = gcc.cc; })
+    else
+      llvmPackages.stdenv;
+  # buildEnv =
+  #   if (stdenv.targetPlatform.isWasm) then
+  #     llvmPackages.stdenv
+  #   else
+  #     overrideCC llvmPackages.stdenv (llvmPackages.clang.override { gccForLibs = gcc11.cc; });
```
Comment (phated) [inline barretenberg.nix]:
```suggestion
```
Cleanup this comment if I get everything working

File: barretenberg.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+in
+buildEnv.mkDerivation
+{
+  pname = "libbarretenberg";
+  version = "0.1.0";
+
+  src = ./cpp;
+
+  nativeBuildInputs = [ cmake ninja ]
+    ++ optionals targetPlatform.isWasm [ binaryen ];
```
Comment (phated) [inline barretenberg.nix:22]: Binaryen fails on aarch64-linux for some reason, I couldn't figure it out.

File: .github/workflows/nix.yml (hunk 1 above)
Comment (kobyhallx) [reply .github/workflows/nix.yml]: Releases, manual and nightly?

Comment (phated) [reply .github/workflows/nix.yml]: I can set up manual and cron for now. This project doesn't have releases yet 😬

File: barretenberg.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  toolchain_file = ./cpp/cmake/toolchains/${targetPlatform.system}.cmake;
+in
+buildEnv.mkDerivation
+{
+  pname = "libbarretenberg";
+  version = "0.1.0";
+
+  src = ./cpp;
+
+  nativeBuildInputs = [ cmake ninja ]
```
Comment (kobyhallx) [inline barretenberg.nix:21]: what do we use ninja for?

Comment (phated) [reply barretenberg.nix:21]: Ninja is now the recommended generator to build barretenberg with CMake. The presets use it and I added it to stay consistent with the way that crypto-dev team builds bb.

**Review by kobyhallx — APPROVED** [review]
👍 LGTM

**Comment by dbanks12** [comment]
Looks mostly fine to me, but knowing nothing about Nix, I cannot follow the Nix files at all. Maybe some additional comments in those files would help newbies like me (and everyone else working on bberg)?

And It seems that the CircleCI checks are no longer tracked here in the PR since there is a github workflow. Is there some way to have it track both?

**Comment by phated** [comment]
@dbanks12 the CircleCI runs are still done and shown. If you expand the "Show all checks" and scroll down you'll see the cci checks:
Screenshot 2023-03-14 at 8 29 53 AM

**Pull request merged**